examples/self_sustaining_parallel.py

```diff
@@ -0,0 +1,205 @@
+previously evaluated are used to choose the hyper-parameters that will be
+explored in the next rounds.
+
+"""
+
+
+##############################################################################
+# The following class implements the computational model. It can be easily
+# adapted to various usecases, since the functions that define the initial
+# tasks and the feedback loop are abstracted
+
+import joblib
+from queue import SimpleQueue
+
+
+class SelfSustainingParallelWork:
+
+    def __init__(self, parallel_n_jobs, parallel_batch_size=1):
+        self.parallel = joblib.Parallel(
+            n_jobs=parallel_n_jobs,
+            return_as="generator_unordered",
+            batch_size=parallel_batch_size,
+        )
+
+        if self.parallel.batch_size == "auto":
+            # Adaptive batch sizing might set batch size to a value higher than
+            # `n_jobs` during compute, which would cause a deadlock, since
+            # then `joblib.Parallel` would risk having to wait for more jobs
+            # than the feedback loop can provide.
+            raise ValueError(
+                "Setting joblib.Parallel `batch_size` parameter to 'auto' is "
+                "not allowed in SelfSustainingParallelWork"
+            )
+
+    def __call__(
+            self,
+            starting_tasks,
+            task_from_output_fn,
+            task_completion_callback
+    ):
+        """Main function that starts the compute and the feedback loop.
+
+        Parameters
+        ----------
+        starting_tasks: iterable of functions
+            An iterable of functions that take no arguments. The compute will
+            start by dispatching those tasks to the workers.
+
+        task_from_output_fn: function
+            A function that the feedback loop will used to create new tasks
+            from previous results. The function must expect one argument, which
+            will be passed the output returned by the latest completed task. It
+            must return either a joblib.delayed-like function, either None, in
+            which case no new task will be dispatched.
+
+        task_completion_callback: function
+            A function that will be called in the calling thread each time an
+            output is retrieved, taking said output as argument. It can be used
+            for recording, plotting, logging,...
+        """
+        # It's required to use a multi-threaded queue here, so that the
+        # callback threads in joblib.Parallel internal mechanisms can also
+        # get the new tasks as soon as it's available and dispatch it.
+        task_feedback_queue = SimpleQueue()
+
+        # Start the compute...
+        input_generator = self._input_generator(
+            starting_tasks, task_from_output_fn, task_feedback_queue
+        )
+        _output_generator = self.parallel(input_generator)
+
+        # ... start retrieving outputs...
+        for output in _output_generator:
```

Comment:
Iterating on the outputs must be done in the calling thread, to benefit from joblib.Parallel safe and elegant termination logic and error reporting. From there the new tasks can be sent to the intput generator using queues.

Implementing it the other way around (iterating on the outputs in the input generator and sending through the queue from the input generator to the calling thread) can work too but it requires tedious error management code and there's no particular gain I can think of.